### PR TITLE
Jlc-pcb repo for pcb-stackup, pcb-rules, pcb-routing-structure, pcb-differential-structure

### PR DESCRIPTION
**Pull request by lchao-jitx**
Initial version for jlc-pcb.

I have created the folder structure as suggested in the Fabrication PRD.
```
jlc-pcb/
  src/
    stackups/
        JLC04161H-7628.stanza
        ...
        JLC06161H-1080.stanza
    rules/
        basic.stanza
        advanced.stanza
```

I have copied the sample `pcb-stackup` and `pcb-routing-structure` and `pcb-differential-routing-structure` for ` JLC04161H-7628.stanza` from the  Fabrication PRD.

I have copied the content of `rules/basic.stanza` from `jlcpcb-basic-rules` in `open-components-database/manufacturers/rules.stanza` and copied the content of `rules/advanced.stanza` from the`jlcpcb-rules` in  `open-components-database/manufacturers/rules.stanza`.

I have added one test for the JLC04161H-7628 stackup in `tests/jlc-pcb.stanza`.

I have the following questions:
Q1. I assume I should change the thickness of layers in  `JLC04161H-7628.stanza` to match the thickness of layers from the page https://jlcpcb.com/impedance 
image
 with these settings:
 
image
   Please confirm.
   Update: I have done that.
    => Closed
Q2. Do we add one stanza file for each of the other 4-layer stackups on that page? Or do we add only a couple, such as the first two 7628 and 3313. 
Here are what they have:
> 1) No requirement Stackup
> 2) JLC04161H-7628 Stackup
> 3) JLC04161H-3313 Stackup
> 4) JLC04161H-1080 Stackup
> 5) JLC04161H-7628A Stackup
> 6) JLC04161H-7628B Stackup
> 7) ...
> 8) ..
> 9) ...
> 10) ...
> 11) JLC04161H-2116C Stackup

   Update: I have added one stanza file for each stackup with 1.6mm thickness for the 4-layer and 6-layer stackups with 1oz and 0.1oz for copper weights, meaning JLC04161H-* and JLC06161H-*. 
          The trace width and spacing for `pcb-routing-structure` and `pcb-differential-routing-structure `for each stackup is modified according to the Impedance Calculator,
   => Closed

Q3. Do I add deftest for  `pcb-routing-structure`, `pcb-differential-routing-strucuture`, and  `pcb-rules`? If so, what to test for?
    Update: I have added basic tests. But print-def() does not work for them. There are no introspect functions for `pcb-routing-structure` and `pcb-differential-routing-strucuture` that we can use either.
  => Will rephase.
Q4. For `pcb-routing-structure` and `pcb-differential-routing-strucuture`, where do I find the parameters to use?
  Update: Erwin helped me to find the info. I have verified the numbers used in `JLC04161H-7628.stanza` and added comments to indicate where I got the numbers.
  => Closed.
Q5. The Fabrication PRD mentioned [PCB impedance calculator](https://jlcpcb.com/pcb-impedance-calculator). Do I use it to calculate some parameters?
  Update: I have used this calculate to get the numbers for Q4.
  => Closed.
Q6. For `pcb-rules`, the file in OCDB refers to this page https://jlcpcb.com/capabilities/pcb-assembly-capabilities, do I get the parameters from there?
  Update: Erwin helped me to find the corresponding numbers from that page.
  => Closed.

======= Additional questions ========
Q7: `print-def()` does not work for `pcb-routing-structure` and `pcb-differential-routing-strucuture` . There are no introspect functions for `pcb-routing-structure` and `pcb-differential-routing-strucuture` we cannot add deftest for them. 

Q8: Where do I find the parameters for `rules/advanced.stanza`? Bhusan looked at the JLCPCB.com site and he could not find out either. What do you have in mind?

Q9: `print-def` does work for `pcb-stackup`, but there is no introspect functions, Thus, it is hard to verify the layers in `deftest`. Perhaps, we should support a function to extract `LayerStack` from a `pck-stackup`.

### CODE REVIEW

File: slm.toml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+name = "jlc-pcb"
+version = "0.1.0"
+[dependencies]
+jsl = { git = "JITx-Inc/jsl", version = "0.3.2" }
```
Comment (bhusang) [inline slm.toml:4]: This version is outdated. We are now on 0.4.0

File: src/main.stanza [hunk 2]
```diff
@@ -0,0 +1,8 @@
+
+defpackage jlcpcb/main:
```
Comment (callendorph) [inline src/main.stanza:2]: Please remove this file.

File: src/rules/basic.stanza [hunk 3]
```diff
@@ -0,0 +1,24 @@
+defpackage jlc-pcb/basic-rules
```
Comment (callendorph) [inline src/rules/basic.stanza:1]: This doesn't compile - you don't have a `:` here. 

You are also missing the `#use-added-syntax(jitx)`
Same for `advanced.stanza`

File: src/rules/basic.stanza [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+public pcb-rules rules :
```
Comment (callendorph) [inline src/rules/basic.stanza:3]: I think we should use their nomenclature here. 

There is rules based on copper weight:

1.  1oz Copper
2.  2oz Copper

Then within `1oz` copper - they separate out by number of layers. 

We can either define separate objects for each rule set. 
Or we can have arguments - but we need to handle all of these cases.

File: src/rules/basic.stanza [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  min-copper-width = 0.127 ; 3.5mil ; 0.127mm for 1 or 2-copper-layer designs
+  min-copper-copper-space = 0.127 ; 3.5mil ; 0.127mm for 1 or 2-copper-layer designs
+  min-copper-hole-space = 0.254 ; 10mil
```
Comment (callendorph) [inline src/rules/basic.stanza:6]: Their rules seem to mostly be in mm - lets just make a comment at the top of the page that everything is in mm, and not try to dual dimension everything.

File: src/stackups/JLC04161H-7628.stanza [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      [copper-35mm core-FR4]
+    ]
+
+public val stackup = create-pcb-stackup(stack)
+
+#for ( 
+  se-name in [SE-50, SE-75]
+  width-val in [0.3424, 0.1311]
+  ):
+  public pcb-routing-structure se-name :
```
Comment (callendorph) [inline src/stackups/JLC04161H-7628.stanza:24]: We will probably want the `clearance` to be a user passable argument with a reasonable default value. 

Applies to both SE and differential

File: src/stackups/JLC04161H-7628.stanza [hunk 7]
```diff
@@ -0,0 +1,55 @@
+#use-added-syntax(jitx)
+defpackage jlc-pcb/JLC04161H-7628:
+  import core
+  import jitx
+  import jsl/layerstack
+  
+val copper-35mm = Copper(0.035, name = "cu")
```
Comment (callendorph) [inline src/stackups/JLC04161H-7628.stanza:7]: I think you mean `um` here - millimeter would not make sense.

File: src/stackups/JLC04161H-7628.stanza [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+val copper-75mm = Copper(0.070, name = "cu75")
```
Comment (callendorph) [inline src/stackups/JLC04161H-7628.stanza:8]: The board I'm seeing [here](https://jlcpcb.com/impedance) for `JLC04161H-7628` is `0.0152` or `15um` not `70um`

File: src/stackups/JLC04161H-7628.stanza [hunk 9]
```diff
@@ -0,0 +1,55 @@
+#use-added-syntax(jitx)
+defpackage jlc-pcb/JLC04161H-7628:
+  import core
+  import jitx
+  import jsl/layerstack
+  
+val copper-35mm = Copper(0.035, name = "cu")
+val copper-75mm = Copper(0.070, name = "cu75") 
+val core-FR4 = FR4(1.065, name = "core")
+val prepreg-FR4 = FR4(0.2104, name = "7628x1")
```
Comment (callendorph) [inline src/stackups/JLC04161H-7628.stanza:10]: JLCPCB lists a dielectric constant - for these layer types. Let's add that definition here as well.

File: src/stackups/JLC04161H-7628.stanza [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    layer(Top):
+      trace-width = width-val
+      clearance = 0.508
+      velocity = 0.19e12
+      insertion-loss = 0.004
+    layer(Bottom):
+      trace-width = width-val
+      clearance = 0.508
+      velocity = 0.19e12
+      insertion-loss = 0.004
```
Comment (callendorph) [inline src/stackups/JLC04161H-7628.stanza:34]: Where are these values coming from ? 

If they are just placeholders then we need to parameterize them somehow. Either as args with defaults or something.

File: stanza.proj [hunk 11]
```diff
@@ -0,0 +1,14 @@
+
+;packages jlc-pcb/* defined-in "./src/"
+;packages jlc-pcb/tests/* defined-in "./tests/"
+
+package jlc-pcb/JLC04161H-7628 defined-in "./src/stackups/JLC04161H-7628.stanza"
```
Comment (callendorph) [inline stanza.proj:5]: Please use glob patterns here:

For example:
```
package jlc-pcb/* defined in "./src/"
```

File: src/main.stanza (hunk 2 above)
Comment (lchao-jitx) [reply src/main.stanza:2]: Done.

File: src/stackups/JLC04161H-7628.stanza (hunk 7 above)
Comment (lchao-jitx) [reply src/stackups/JLC04161H-7628.stanza:7]: Changed.

File: slm.toml (hunk 1 above)
Comment (lchao-jitx) [reply slm.toml:4]: Changed. Thanks.

File: src/stackups/JLC04161H-7628.stanza (hunk 9 above)
Comment (lchao-jitx) [reply src/stackups/JLC04161H-7628.stanza]: I have added the dialectric constant for the prepreg layers in the function FR4 (type:PrepregType, thickness:Double, name:String) in `shared.stanza`
```
public defn FR4 (type:PrepregType, thickness:Double, name:String) -> LayerSpec :
  val material = switch {type == _} :
    \|7628| : FR4-Material-7628
    \|3313| : FR4-Material-3313
    \|1080| : FR4-Material-1080
    \|2116| : FR4-Material-2116
  FR4(thickness, material, name = name)
```
The dialectric constant inserted into `FR4-Material-7628` through the function `dialectric-coefficient (type:PrepregType)`
```
defn dialectric-coefficient (type:PrepregType) -> Double :
  switch {type == _} :
    \|7628| : 4.4
    \|3313| : 4.1
    \|1080| : 3.91
    \|2116| : 4.16

public val FR4-Material-7628 = DielectricMaterial(
    name = "FR4 (7628)"
    description = "FR4 for JLC 7628"
    dielectric-coefficient = dialectric-coefficient(\|7628|) ; @ 1GHz
    loss-tangent = 0.25 ; @ 1GHz
  )
```

File: stanza.proj (hunk 11 above)
Comment (lchao-jitx) [reply stanza.proj]: Changed to:
```
packages jlc-pcb/* defined-in "./src/"
packages jlc-pcb/tests/* defined-in "./tests/"
```

File: src/stackups/JLC04161H-7628.stanza (hunk 8 above)
Comment (lchao-jitx) [reply src/stackups/JLC04161H-7628.stanza]: I have made changes to use the parameters from the jlcpcb.com website.

Per @Bhusan's request,
I have also added function `create-pcb-stackup-from-name` to create all 4-layer stackups by retrieving thickness, copper weights from the encoded JPC-PCB stackup names.

Comment (lchao-jitx) [reply src/stackups/JLC04161H-7628.stanza]: I wonder how we can write deftest for pcb-stackup?

Except `print-def(stackup)`, I don't know of any function to extract info from a pcb-stackup JJITXDef. Maybe we should add some introspect functions?

Perhaps, we should have a function to extract LayerStack from a pcb-stackup. Then, we can compare the extracted LayerStack.

Also, I thought I could compare the output of `print-def(stackup)`. But when I re-direct stdout to a string for  `print-def(stackup)`, I would get a serialization error.

File: src/rules/basic.stanza (hunk 3 above)
Comment (lchao-jitx) [reply src/rules/basic.stanza]: Fixed.

**Comment by lchao-jitx** [comment]
In the last commit "[defstruct and functions for pcb-rules](https://github.com/JITx-Inc/jlc-pcb/pull/1/commits/4d8fa1bf1eb6ad1cd874a5afd6b9fda4b74da17c)", I added defstruct and functions to follow the model of LayerStack and pcb-stackup.

Is that what you have in mind?

File: src/rules/basic.stanza (hunk 4 above)
Comment (lchao-jitx) [reply src/rules/basic.stanza]: I see that these are the ones that vary with copper weight or the number of layers:
```
    "Minimum Annular Ring" => to-hashtable<String,?> $ 
      [
        "Minimum Annular Ring" => [
          "1oz Copper" => 0.13 ; 0.13mm
          "2oz Copper" => 0.2  ; 0.2mm
        ]
        "PTH" => [
          "1oz Copper" => 0.3 ; 0.3mm
          "2oz Copper" => 0.3  ; 0.3mm
        ]
      ]
    "Minimum trace width and spacing" => to-hashtable<String,?> $ 
      [
        "Min. Trace width" => to-hashtable<String,?> $ [
          "1-2 Layers" => 0.127 ; 0.127mm
          "4-6 Layers" => 0.09  ; 0.09mm
          "2oz Copper weight" => 0.2 ; 0.2mm
        ]
        "Min. Spacing" => to-hashtable<String,?> $ [
          "1-2 Layers" => 0.127 ; 0.127mm
          "4-6 Layers" => 0.09  ; 0.09mm
          "2oz Copper weight" => 0.2 ; 0.2mm
        ]
      ]
    "BGA" => to-hashtable<String,?> $ 
      [
        "Min. BGA Pad Dimensions" => to-hashtable<String,?> $ [
          "1/2 Layers" => 0.25 ; 0.25mm
          "4/6 Layers" => 0.25 ; 0.25mm
        ]
        "Min. Distance Between BGA" => to-hashtable<String,?> $ [
          "1/2 Layers" => 0.127 ; 0.127mm
          "4/6 Layers" => 0.127 ; 0.127mm
        ]
      ]
```
See `src/rules/shared.stanza` for all of the rules from https://jlcpcb.com/capabilities/pcb-capabilities

File: src/rules/basic.stanza (hunk 5 above)
Comment (lchao-jitx) [reply src/rules/basic.stanza]: Done.

File: src/stackups/JLC04161H-7628.stanza (hunk 6 above)
Comment (lchao-jitx) [reply src/stackups/JLC04161H-7628.stanza]: I have added `clearance` as an optional argument for `pcb-routing-structure` and `pcb-differential-routing-structure` in `src/stackups/JLC04161H-7628.stanza`:
```
doc:\<DOC>
  The parameters for the pcb-routing-structure and pcb-differential-routing-structure
  are retrieved from the PCB Impedance Calculator
   - For Coplanar Single-Ended, the valid values for Impedance is from 20 to 90.
   - For Coplanar Differential Pair, the valid values for Impedance is from 50 to 150
   - `trace-width` comes from the "Trace Width" column of the Impedance Calculator
   - `pair-spacing` comes from the "Trace Spacing" column of the Impedance Calculator
   - `clearance` comes from the "Impedance trace to copper" column of the Impedance Calculator
   - The function `velocity` calculate the velocity from the dielectric constant of the material for "7628"

  Example Usage:
    To use the default clearance: jlc-pcb/JLC04161H-7628/SE-50()
    To use the user-given clearance: jlc-pcb/JLC04161H-7628/SE-50(clearance = 0.5)

   Reference: https://jlcpcb.com/pcb-impedance-calculator
<DOC>

; For Coplanar Single-Ended
;  - width-val comes from the "Trace Width" column
#for ( 
  se-name in [SE-50, SE-75]
  width-val in [0.3424, 0.1311]
  ):
  public pcb-routing-structure se-name (
      -- clearance:Double = 0.508) :
    layer(Top):
      trace-width = width-val
      clearance = clearance
      velocity = velocity("7628")
      insertion-loss = 0.004
    layer(Bottom):
      trace-width = width-val
      clearance = 0.508
      velocity = velocity("7628")
      insertion-loss = 0.004
```

If this looks ok, I'll make the same change to the other files.

File: src/stackups/JLC04161H-7628.stanza (hunk 10 above)
Comment (lchao-jitx) [reply src/stackups/JLC04161H-7628.stanza]: Erwin told me how to calculate the velocity. So, I have added a function to calculate it in `src/stackups/shared.stanza` from the dielectic constant.
```
doc:\<DOC>
    Calculate the signal propagation velocity 

    @param epsilon_r (dielectric-constant): The relative permittivity of the material. Default is 4.5.

    @return The signal velocity in the FR4 material (in mm/second).
<DOC>
defn calculate_velocity (epsilon_r = 4.5) :
    ;Speed of light in vacuum (mm/s)
    val light-speed = 3.0e11
    ;Calculate the signal velocity
    val velocity = light-speed * (math/pow(epsilon_r -0.5)) ; equivalent to light-speed / (math/pow(epsilon_r 0.5)
    velocity

doc:\<DOC>
    Calculate the signal propagation velocity in FR4 material.

    @param type: The material type: "core" or "7628", "3313", "1080" or "2116"
    @return The signal velocity of the dielectric material (in mm/second).
<DOC>
public defn velocity (type:String) :
  calculate_velocity(dielectric-constant(type))
```

I don't know where `insertion-loss` is coming from. So, leaving it as is.

Comment (callendorph) [reply src/stackups/JLC04161H-7628.stanza]: In the `jsl/si/physics` package there are defined functions for computing various quantities like this.

**Comment by lchao-jitx** [comment]
Done. I have changed it to use `jsl/si/physics/phase-velocity`.

File: src/stackups/JLC04161H-7628.stanza (hunk 10 above)
Comment (lchao-jitx) [reply src/stackups/JLC04161H-7628.stanza]: Done. I have changed it to use `jsl/si/physics/phase-velocity`.

File: slm.toml [hunk 12]
```diff
@@ -0,0 +1,4 @@
+name = "jlc-pcb"
+version = "0.4.0"
```
Comment (callendorph) [inline slm.toml:2]: Let's use `0.1.0` as this will be the first release.

File: src/rules/basic.stanza (hunk 4 above)
Comment (lchao-jitx) [reply src/rules/basic.stanza]: I have added functions to retrieve the copper trace width/spacing based on the number of layers or cooper weight of a given stackup.

I am not sure if I interpreted it appropriately. Here is the code added:
```
defn retrieve-min-trace-width (stackup-prop:StackupProperty, rules-table:HashTable<String,?>) -> Double :
  val trace-width-rules-section = rules-table["Minimum trace width and spacing"]["Min. Trace width"]
  if num-layers(stackup-prop) <= 2 :
    trace-width-rules-section["1-2 Layers"]
  else if num-layers(stackup-prop) <= 6 :
    trace-width-rules-section["4-6 Layers"]
  else : ; 7 or more layers
    if outer-copper-weight(stackup-prop) ==  \|2oz| or inner-copper-weight(stackup-prop) ==  \|2oz| :
      trace-width-rules-section["2oz Copper weight"]
    else :
      minimum $ values $ trace-width-rules-section
```

Here is the function to extract the values for annular ring:
```
defn retrieve-annular-ring-rules (rules-table:HashTable<String,?>) -> AnnularRingRules :
  val ring-rules-section = rules-table["Minimum Annular Ring"]
  AnnularRingRules(min-annular-ring, pth) where :
    val min-annular-ring = to-hashtable<CopperWeightType,Double> $
      [ \|1oz| => ring-rules-section["Minimum Annular Ring"]["1oz Copper"]
        \|2oz| => ring-rules-section["Minimum Annular Ring"]["2oz Copper"]
      ]
    val pth = to-hashtable<CopperWeightType,Double> $
      [ \|1oz| => ring-rules-section["PTH"]["1oz Copper"]
        \|2oz| => ring-rules-section["PTH"]["2oz Copper"]
      ]
defstruct AnnularRingRules :
  min-annular-ring:HashTable<CopperWeightType,Double>
  pth:HashTable<CopperWeightType,Double>
```

Tests are added.
```
; Test Rules for 2oz stackcup - use 8-layer stackup so a different trace width is used
deftest test-rules-for-JLC02161H-7628-from-name :
  val test-base = create-rules-for-JLC02161H-7628()
  val test-rules = jlc-pcb/rules/basic/create-pcb-rules-from-name("JLC02161H-7628")
  #EXPECT(empty? $ diff-clearances(clearances(test-rules), clearances(test-base)))

; Test Rules for 2oz stackcup - use 8-layer stackup so a different trace width is used
deftest test-rules-for-JLC08162H-7628-from-name :
  val test-base = create-rules-for-JLC08162H-7628()
  val test-rules = jlc-pcb/rules/basic/create-pcb-rules-from-name("JLC08162H-7628")
  #EXPECT(empty? $ diff-clearances(clearances(test-rules), clearances(test-base)))
```

File: src/stackups/JLC04161H-7628.stanza (hunk 10 above)
Comment (lchao-jitx) [reply src/stackups/JLC04161H-7628.stanza]: I have added both `clearance` and `insertion-loss`  as optional arguments with default values:

```
#for ( 
  se-name in [SE-50, SE-75]
  width-val in [0.3424, 0.1311]
  ):
  public pcb-routing-structure se-name (
      -- clearance:Double = 0.508
         insertion-loss:Double = 0.004) :
    layer(Top):
      trace-width = width-val 
      clearance = clearance
      velocity = velocity("7628")
      insertion-loss = insertion-loss
    layer(Bottom):
      trace-width = width-val
      clearance = clearance
      velocity = velocity("7628")
      insertion-loss = insertion-loss
``` 

This is done only in `src/stackups/JLC04161H-7628.stanza` now.
Once we finalize it, I'll duplicate the structure to other files.

Comment (lchao-jitx) [reply src/stackups/JLC04161H-7628.stanza]: The content of `pcb-routing-structure` is almost identical.
Should I create a function like `create-routing-structure` with a list of parameters to create the pcb-routing-structures?

**Pull request closed without merging**